### PR TITLE
Add serve external kv store

### PR DESCRIPTION
**Pull request by jiaodong**
## Why are these changes needed?

In order to achieve cluster level fault tolerance, we need to be able to checkpoint controller states to external storage rather than GCS, which fate share with ray cluster.

## Changes made
 - Moved kv store related files and tests into separate folder, and changed related imports 
 - Added two new KV store classes 
 - Added ray serve KV store interface 

## Test plan

```
❯ pytest test_kv_store.py::test_ray_serve_external_kv_local_disk
Test session starts (platform: darwin, Python 3.8.10, pytest 5.4.3, pytest-sugar 0.9.4)
rootdir: /Users/jiaodong/Workspace/ray/python
plugins: sugar-0.9.4, asyncio-0.15.1, timeout-1.4.2, lazy-fixture-0.6.3, rerunfailures-10.0
collecting ...
 ray/serve/tests/test_kv_store.py ✓

❯ pytest test_kv_store.py::test_ray_serve_external_kv_aws_s3
Test session starts (platform: darwin, Python 3.8.10, pytest 5.4.3, pytest-sugar 0.9.4)
rootdir: /Users/jiaodong/Workspace/ray/python
plugins: sugar-0.9.4, asyncio-0.15.1, timeout-1.4.2, lazy-fixture-0.6.3, rerunfailures-10.0
collecting ...
 ray/serve/tests/test_kv_store.py ✓       
```                                                                                                                                                                                            

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/kv_store.py
Comment (edoakes) [inline python/ray/serve/kv_store.py:87]: Hmm the local_mode flag is kind of weird given that it uses a completely separate code path for every operation, can we just make these two separate stores? `LocalDiskKVStore` and `S3KVStore`

File: python/ray/serve/kv_store.py
Comment (edoakes) [inline python/ray/serve/kv_store.py:65]: We should add a common base class to define the interface for all of the impls here. You can use an abstract base class for this:
https://www.geeksforgeeks.org/abstract-classes-in-python/

File: python/ray/serve/kv_store.py
Comment (edoakes) [inline python/ray/serve/kv_store.py:119]: Can you explain the file layout and semantics for this in a comment? This should be the docstring for the `LocalDiscKVStore`

File: python/ray/serve/tests/test_kv_store.py
Comment (edoakes) [inline python/ray/serve/tests/test_kv_store.py:66]: We should probably also run some regular integration tests with this kv store. Maybe the controller failure test?

**Review by edoakes — CHANGES_REQUESTED** [review]

File: python/ray/serve/tests/test_kv_store.py
Comment (jiaodong) [reply python/ray/serve/tests/test_kv_store.py:66]: Sure but a simpler version given current one didn't fully pass yet ...

File: python/ray/serve/storage/tests/test_kv_store.py
Comment (edoakes) [inline python/ray/serve/storage/tests/test_kv_store.py:15]: FYI you'll need to add this to the bazel BUILD file in serve root for it to run in CI

**Comment by edoakes** [comment]
@jiaodong looks like there's a merge conflict. Is this ready to review again otherwise?

File: python/ray/serve/storage/kv_store.py
Comment (simon-mo) [inline python/ray/serve/storage/kv_store.py:102]: instead of inventing our own format, can we please just write it to sqlite? https://docs.python.org/3/library/sqlite3.html

**Comment by jiaodong** [comment]
> @jiaodong looks like there's a merge conflict. Is this ready to review again otherwise?

yep it is, i have it running fine locally (might need more config to let it run automatically tho, plus aws credential for unit test)

It's written on master ~25 hrs ago .. let me rebase on master and re-submit.

File: python/ray/serve/storage/kv_store.py
Comment (jiaodong) [reply python/ray/serve/storage/kv_store.py:102]: that's my first choice as well but i didn't find it anywhere else in ray oss import or testing ... to me it looks like we haven't made sqlite a dependency and ideally not adding it for simple case like this ?

Comment (simon-mo) [reply python/ray/serve/storage/kv_store.py:102]: It's part of python stdlib

Comment (simon-mo) [reply python/ray/serve/storage/kv_store.py:102]: You can see past example here https://github.com/ray-project/ray/pull/5937/files and find kv_store_service

File: python/ray/serve/storage/tests/test_kv_store.py
Comment (jiaodong) [inline python/ray/serve/storage/tests/test_kv_store.py:88]: @edoakes  @simon-mo do we have any best practice to handle this in CI/CD ?

Comment (simon-mo) [reply python/ray/serve/storage/tests/test_kv_store.py:88]: Not yet. Let's just test with file system for now. You can skip this test.

File: dashboard/modules/snapshot/snapshot_head.py
Comment (edoakes) [inline dashboard/modules/snapshot/snapshot_head.py:103]: why change this to cloudpickle? This needs to be json serializable anyways in order to return it in an HTTP response

File: python/ray/serve/controller.py
Comment (edoakes) [inline python/ray/serve/controller.py:148]: ditto here we should probably keep this JSON

File: python/ray/serve/storage/kv_store.py
Comment (edoakes) [inline python/ray/serve/storage/kv_store.py:182]: will these be read from the filesystem  / env variables if not provided?

File: python/ray/serve/storage/tests/test_kv_store.py
Comment (edoakes) [inline python/ray/serve/storage/tests/test_kv_store.py:80]: are these still relevant now that we're using sqlite3?

File: python/ray/serve/storage/tests/test_kv_store.py
Comment (edoakes) [inline python/ray/serve/storage/tests/test_kv_store.py:72]: no need for the `ray_serve` prefix

File: python/ray/serve/storage/tests/test_kv_store.py
Comment (edoakes) [reply python/ray/serve/storage/tests/test_kv_store.py:88]: +1, let's skip the S3 test for now to get this merged. @simon-mo can you follow up with @jiaodong on how to test it in CI though?

File: python/ray/serve/storage/tests/test_kv_store.py
Comment (edoakes) [inline python/ray/serve/storage/tests/test_kv_store.py:7]: why do we import `shared_serve_instance`? I don't think it's used.

File: python/ray/serve/tests/conftest.py
Comment (edoakes) [inline python/ray/serve/tests/conftest.py:17]: what's the context for this change?

File: python/ray/serve/storage/kv_storage_base.py
Comment (edoakes) [inline python/ray/serve/storage/kv_storage_base.py:1]: nit: can you rename this `kv_store_base.py` to match the other filename (`kv_store.py`)?

**Review by edoakes — COMMENTED** [review]
Looks good mostly just some nits. Tests are failing due to lack of S3 credentials, let's skip that test to get this PR merged and we can enable it separately.

File: dashboard/modules/snapshot/snapshot_head.py
Comment (jiaodong) [reply dashboard/modules/snapshot/snapshot_head.py:103]: this is to keep ray serve kvstore only deal with bytes type since json.dumps / loads deal with string, last week we had #17489 that changed it to Union[str, bytes].

File: python/ray/serve/storage/kv_store.py
Comment (jiaodong) [reply python/ray/serve/storage/kv_store.py:182]: i left it to user code (in our case, unit test input) to decide but it works either way

File: python/ray/serve/storage/tests/test_kv_store.py
Comment (jiaodong) [reply python/ray/serve/storage/tests/test_kv_store.py:80]: nope, should remove :)

File: python/ray/serve/storage/tests/test_kv_store.py
Comment (jiaodong) [reply python/ray/serve/storage/tests/test_kv_store.py:7]: test will failing complaining `shared_serve_instance` not imported if it's not there

File: python/ray/serve/tests/conftest.py
Comment (jiaodong) [reply python/ray/serve/tests/conftest.py:17]: ditto, i found it that we need to import both `serve_instance` and `shared_serve_instance` in order to let test pass

File: python/ray/serve/storage/tests/test_kv_store.py
Comment (edoakes) [reply python/ray/serve/storage/tests/test_kv_store.py:7]: huh so weird. pytest works in mysterious ways

Comment (simon-mo) [reply python/ray/serve/storage/tests/test_kv_store.py:7]: Hmmm they should not be imported! If it fails with import error make sure serve instance is an argument name for the test function first. So that the serve_instance variable comes from function argument not directly import.

File: python/ray/serve/tests/conftest.py
Comment (simon-mo) [reply python/ray/serve/tests/conftest.py:17]: That shouldn't be the case, please revert this so we are making sure the test is setup properly.

**Comment by simon-mo** [comment]
ah actually i think the source of the testing import issue is because you put the test inside python/ray/serve/storage/tests/test_kv_store.py , you can just move it to python/ray/serve/tests/test_kv_store.py  and your import issue should be gone

File: python/ray/serve/BUILD
Comment (simon-mo) [inline python/ray/serve/BUILD:154]:
```suggestion
    srcs = serve_storage_tests_srcs + serve_test_srcs,
```

File: python/ray/serve/storage/kv_store.py
Comment (architkulkarni) [inline python/ray/serve/storage/kv_store.py:13]: Is there any simple way to keep this in sync with the implementation of this function in the different Serve KV store classes?  If we change this format in the future, we'll have to remember to update it in multiple places, but maybe it's not a big deal.

Comment (jiaodong) [reply python/ray/serve/storage/kv_store.py:13]: i think we should keep single source of truth on this as well, and based on your PR it seems we have the need to get storage key without an kv store instance. In this case i think we should just keep this function as the only one for imports like this, and for other kv store types, its interface requires implementation of get_storage_key so we can expect it to be consistent as long as it's using the same class.

**Review by edoakes — APPROVED** [review]
Unblocking but leaving final review to @simon-mo

File: dashboard/modules/snapshot/snapshot_head.py
Comment (wuisawesome) [inline dashboard/modules/snapshot/snapshot_head.py:101]:
```suggestion
        val_bytes = _internal_kv_get(key, client) or "{}".encode("utf-8")
```
?

Comment (jiaodong) [reply dashboard/modules/snapshot/snapshot_head.py:101]: ah good call, it's empty dict from its source in controller

**Review by wuisawesome — APPROVED** [review]
Also approving to unblock since the snapshot stuff looks trivial, but leaving it up to @simon-mo

File: python/ray/serve/storage/kv_store.py
Comment (simon-mo) [inline python/ray/serve/storage/kv_store.py:7]:
```suggestion
import sqlite3

import boto3
from botocore.exceptions import ClientError

import ray.experimental.internal_kv as ray_kv
from ray.serve.storage.kv_store_base import KVStoreBase
from ray.serve.utils import logger
```
ray codebase follow the import order of
```
import stdlib
# new line
import third party lib
# new line
import ray stuff 
```

Comment (simon-mo) [reply python/ray/serve/storage/kv_store.py:7]: additionally, we need to gate the boto3 import because Ray Serve users might not have it installed and it's not part of the dependency. So you have to either import it conditionally or import it within the S3 key value store path. The last restore would be add it to setup.py as a Serve dep.

File: python/ray/serve/storage/kv_store.py
Comment (simon-mo) [inline python/ray/serve/storage/kv_store.py:10]: type annotate

Comment (simon-mo) [reply python/ray/serve/storage/kv_store.py:10]: also from snapshot_head, it's called with multiple argument `get_storage_key(SERVE_CONTROLLER_NAME, SERVE_SNAPSHOT_KEY)` so it might error there?

**Review by simon-mo — CHANGES_REQUESTED** [review]
I can't find python/ray/serve/tests/storage/test_kv_store.py in this diff

**Comment by simon-mo** [comment]
also merge conflict :( sorry

File: python/ray/serve/storage/kv_store.py
Comment (jiaodong) [reply python/ray/serve/storage/kv_store.py:10]: ah good call i forgot to change this while rebasing

**Review by simon-mo — APPROVED** [review]

**Comment by simon-mo** [comment]
https://buildkite.com/ray-project/ray-builders-pr/builds/11075#e58e1f87-6009-4f16-b9e6-a3b61a1e4b40
```
E           types.RayTaskError(ModuleNotFoundError): ray::test_import() (pid=28995, ip=172.18.0.3)
--
  | E             File "/ray/python/ray/tests/test_runtime_env_complicated.py", line 737, in test_import
  | E               from ray import serve  # noqa
  | E             File "/tmp/ray/session_2021-08-11_05-41-18_557612_28218/runtime_resources/conda/ray-399ee03a743383a88f4433896502239846584941/lib/python3.6/site-packages/ray/serve/__init__.py", line 1, in <module>
  | E               from ray.serve.api import (connect, start, get_replica_context, get_handle,
  | E             File "/tmp/ray/session_2021-08-11_05-41-18_557612_28218/runtime_resources/conda/ray-399ee03a743383a88f4433896502239846584941/lib/python3.6/site-packages/ray/serve/api.py", line 26, in <module>
  | E               from ray.serve.controller import ReplicaTag, ServeController
  | E             File "/tmp/ray/session_2021-08-11_05-41-18_557612_28218/runtime_resources/conda/ray-399ee03a743383a88f4433896502239846584941/lib/python3.6/site-packages/ray/serve/controller.py", line 9, in <module>
  | E               from ray.serve.backend_state import BackendState
  | E             File "/tmp/ray/session_2021-08-11_05-41-18_557612_28218/runtime_resources/conda/ray-399ee03a743383a88f4433896502239846584941/lib/python3.6/site-packages/ray/serve/backend_state.py", line 18, in <module>
  | E               from ray.serve.storage.kv_store import RayInternalKVStore
  | E           ModuleNotFoundError: No module named 'ray.serve.storage'
```
You need to add an empty `__init__.py` file as `python/ray/serve/storage/__init__.py` so it's packaged in the ray wheel.

**Pull request merged**
